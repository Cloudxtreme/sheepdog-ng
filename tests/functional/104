#!/bin/bash

# Test random data stream input at the listening port

. ./common

for i in `seq 0 2`; do
    _start_sheep $i
done

_wait_for_sheep 3

_cluster_format -c 2

$DOG vdi create test 100M
_random | $DOG vdi write test &

# Test all zero data input
cat /dev/zero | nc localhost 7000

# Test random data input
for i in `seq 0 9`; do
    cat /dev/urandom | nc localhost 7000
done

# Check an invalid proto_ver
echo -ne '\x02\x01\x00\x00\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x430\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07' | nc localhost 7000

# Check a valid proto_ver and opcode, but the opcode is logically wrong
echo -ne '\x09\x01\x00\x00\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x430\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07' | nc localhost 7000

# Check a valid proto_ver but invalid flags
echo -ne '\x09\x82\x12\x10\x00\x00\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x430\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07' | nc localhost 7000

# Check a valid proto_ver, opcode, flags, but invalid epoch
echo -ne '\x01\x01\x00\x00\x01\x00\x00\x00\x09\x01\x30\x30\x30\x30\x4e\x430\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07' | nc localhost 7000

# Check bad length for create_and_write
echo -ne '\x01\x01\x01\x00\x00\x00\x00\x00\x09\x01\x30\x30\x30\x30\x4e\x430\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07' | nc localhost 7000

# Check create_and_write req with bad oid, object should be created only
echo -ne '\x01\x01\x01\x00\x00\x00\x00\x00\x09\x01\x30\x30\x30\x00\x00\x00\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07\x70123141231231234123151523423423423423424123123123123' | nc localhost 7000

# Check the write req with bad oid, only object not found error
echo -ne '\x01\x03\x01\x00\x00\x00\x00\x00\x09\x01\x30\x30\x30\x00\x00\x00\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07\x70123141231231234123151523423423423423424123123123123' | nc localhost 7000

# Check the read req with bad oid, only object not found error
echo -ne '\x01\x02\x01\x00\x00\x00\x00\x00\x09\x01\x30\x30\x30\x00\x00\x00\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x09\x01\x30\x30\x30\x30\x4e\x43\x01\x02\x03\x04\x05\x06\x07\x70123141231231234123151523423423423423424123123123123' | nc localhost 7000
wait
sync
grep 'bad data stream' $STORE/0/sheep.log | wc -l

$DOG vdi check test
$DOG node list

