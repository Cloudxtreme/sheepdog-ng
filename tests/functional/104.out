QA output created by 104
using backend plain store
16
finish check&repair test
  Id   Host:Port         V-Nodes       Zone
   0   127.0.0.1:7000      	128          0
   1   127.0.0.1:7001      	128          1
   2   127.0.0.1:7002      	128          2
